release/src/router/plc-utils/plc/trafficgen.1
.TH trafficgen 1 "June 2018" "plc-utils-2.1.11" "Qualcomm Atheros Powerline Toolkit"

.SH NAME
trafficgen - Traffic generation utility for 75xx devices

.SH SYNOPSIS
.BR trafficgen
.RI [ options ] 
.RI [ node ] 
.RI [ peer ] 

.SH DESCRIPTION
The Traffic generation utility creates traffic between two given devices node and peer and prints the average transmit and receive PHY rates between them.
Output is column-oriented and suitable for loading into a spreadsheet or being read by a simple program.

.PP
This program is part of the Qualcomm Atheros Powerline Toolkit.
See the \fBplc\fR man page for an overview and installation instructions.

.SH CHIPSETS
QCA75xx 

.SH OPTIONS

.TP
-\fBa \fItraffic_rate\fR
The rate (Mbps) at which traffic is sent to peer device. 
The minimum value is \fB0\fR and the maximum value is \fB50\fR .
If the value is 0, this utility sends traffic at approximately 50Mbps

.TP
.RB - c
Display coded PHY rates.
Coded PHY rates are displayed by default so this option is redundant but it has been kept for backword compatibility.
This option and option \fB-u\fR are mutually exclusive.

.TP
-\fBd \fIseconds\fR
Traffic duration in seconds per connection.
Directional traffic is systematically sent between each pair of powerline network devices for this period of time.
Consequently, the total time needed to generate traffic between two network devices will be approximately 2 times this value, so don't get carried away.
The minimum value is \fB0\fR second and the maximum value is \fB60\fR seconds.

.TP
.RB - e
Redirects stderr messages to stdout.
By convention status and error messages are printed on stderr while primary program output is printed on stdout.
This option prints all output on stdout in cases where this is desired.

.TP
-\fBi \fIinterface\fR
Select the host Ethernet interface.
All requests are sent via this host interface and only reponses received via this host interface are recognized.
The default interface is \fBeth1\fR because most people use \fBeth0\fR as their principle network connection; however, if environment string "PLC" is defined then it takes precedence over the default interface.
This option then takes precedence over either default.

.TP
.RB - n
Reads the average transmit and receive PHY rate for each device using VS_NW_INFO and prints it on stdout.
Each output line displays the ethernet interface, source address, destination address and either the transmit or receive PHY rate.

.TP
-\fBo \fImilliseconds\fR
Read timeout in milliseconds.
Values range from 0 through UINT_MAX.
This is the maximum time allowed for a response.
The default is shown in brackets on the program menu.

.TP
.RB - q
Enter quiet mode.
Progress messages are suppressed.

.TP
.RB - r
Reads device hardware and software revision information using VS_SW_VER and prints it on stdout.

.TP
.RB - R
Resets the device using VS_RS_DEV.

.TP
.RB - t
Generate powerline traffic between the local device and remote device( by using VS_TRAFFIC_GENERATOR MME ) on each powerline network connected to the host.

.TP
.RB - T
Generate unidirectional powerline traffic between the source device and destination device( by using VS_TRAFFIC_GENERATOR MME ). 2 MAC Addresses are specified in the argument , the first is taken as the source and the second one as destination.

.TP
.RB - u
Display un-coded PHY rates.
Coded PHY rates are displayed by default.
This option and option \fB-c\fR are mutually exclusive.

.TP
.RB -v
Enter verbose mode.
All Etherenet frames sent or received by the program are displayed on stdout.

.TP
.RB - x
Exit program on first error with a non-zero exit code.
This option allows shell scripts to detect failed or incomplete operations and take the appropriate action.

.TP
.RB - ? ,-- help
Print program help summary on stdout.
This option takes precedence over other options on the command line.

.TP
.RB - ! ,-- version
Print program version information on stdout.
This option takes precedence over other options on the command line.
Use this option when sending screen dumps to Atheros Technical Support so that they know exactly which version of the Linux Toolkit you are using.

.SH ARGUMENTS 

.TP
.B node
The MAC address of local powerline device.
The default address is \fBlocal\fR.
See \fBDEVICES\fR for information about symbolic device addresses.

.TP
.BR peer
The MAC address of remote powerline device.
A MAC address is a 48-bit value entered as 12 hexadecimal digits in upper, lower or mixed character case.
Octets may be separated with colons for clarity.
For example, "00b052000001", "00:b0:52:00:00:01" and "00b052:000001" are valid and equivalent.

.PP
The following MAC address is special and may be entered by name instead of number.

.TP
.BR local
A synonym for the Qualcomm Atheros vendor specific Local Management Address (LMA), \fB00:B0:52:00:00:01\fR.
All local Atheros devices recognize this address but remote and foreign devices do not.
A remote device is any device at the far end of a powerline connection.
A foreign device is any device not manufactured by Atheros.

.SH REFERENCES
See the Qualcomm Atheros HomePlug AV Firmware Technical Reference Manual for more information.

.SH EXAMPLES
The following example displays the average transmit and receive PHY rates from each of two devices.
One is a local device connected to interface \fBeth2\fR and the other is a remote device at the far end of the powerline.
Observe the rates shown.

.PP
   # trafficgen.exe -i3 -td3 00B052000010 -nu
  nic3 00:B0:52:00:00:99 00:B0:52:00:00:10 RX 1000 mbps MIMO
  nic3 00:B0:52:00:00:99 00:B0:52:00:00:10 TX 1011 mbps MIMO

.SH DISCLAIMER
Atheros HomePlug AV Vendor Specific Management Message Entry structure and content is proprietary to Qualcomm Atheros, Ocala FL USA.
Consequently, public information may not be available.
Qualcomm Atheros reserves the right to modify message structure and content in future firmware releases without any obligation to notify or compensate users of this program.

.SH SEE ALSO
.BR PLC ( 1 ), 
.BR amprate ( 1 ), 
.BR int6krate ( 1 ) 
.BR plcrate ( 1 ) 

.SH CREDITS
 Nisha K <nishk@qti.qualcomm.com>
'
'



release/src/router/plc-utils/plc/trafficgen.c
/*====================================================================*
 *	Copyright (c) 2018-2020 Qualcomm Technologies, Inc.
 *	All Rights Reserved.
 *	Confidential and Proprietary - Qualcomm Technologies, Inc.
 *====================================================================*/
 
/*====================================================================*
 *
 *   signed Traffic3 (struct plc * plc);
 *
 *   generate data traffic between two devices using MME 
 *	 VS_TRAFFIC_GENERATOR
 *
 *   Contributor(s):
 *			Nisha K(nishk@qti.qualcomm.com)
 *--------------------------------------------------------------------*/
/*====================================================================*
 *   system header files;
 *--------------------------------------------------------------------*/

#include <unistd.h>
#include <stdlib.h>
#include <stdint.h>
#include <limits.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/putoptv.h"
#include "../tools/memory.h"
#include "../tools/number.h"
#include "../tools/symbol.h"
#include "../tools/types.h"
#include "../tools/flags.h"
#include "../tools/files.h"
#include "../tools/error.h"
#include "../plc/plc.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../plc/chipset.c"
#include "../plc/Devices.c"
#include "../plc/Confirm.c"
#include "../plc/Display.c"
#include "../plc/Failure.c"
#include "../plc/Request.c"
#include "../plc/ReadMME.c"
#include "../plc/SendMME.c"
#include "../plc/LocalDevices.c"
#include "../plc/PLCSelect.c"
#include "../plc/ResetDevice.c"
#include "../plc/PhyRates2.c"
#include "../plc/Traffic3.c"
#include "../plc/WaitForStart.c"
#endif

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/uintspec.c"
#include "../tools/hexdump.c"
#include "../tools/hexencode.c"
#include "../tools/hexdecode.c"
#include "../tools/todigit.c"
#include "../tools/checkfilename.c"
#include "../tools/checksum32.c"
#include "../tools/error.c"
#include "../tools/fdchecksum32.c"
#include "../tools/synonym.c"
#include "../tools/typename.c"
#endif

#ifndef MAKEFILE
#include "../ether/openchannel.c"
#include "../ether/closechannel.c"
#include "../ether/readpacket.c"
#include "../ether/sendpacket.c"
#include "../ether/channel.c"
#endif

#ifndef MAKEFILE
#include "../mme/EthernetHeader.c"
#include "../mme/QualcommHeader.c"
#include "../mme/QualcommHeader1.c"
#include "../mme/UnwantedMessage.c"
#include "../mme/MMECode.c"
#endif

/*====================================================================*
 *   program constants;
 *--------------------------------------------------------------------*/

#define PLCRATE_WAIT 0
#define PLCRATE_LOOP 1

/*===========================================================================================*
 *
 *   void manager (struct plc * plc, signed count, signed pause, uint8_t trafficgen_rate);
 *
 *   perform operations in logical order despite any order specfied
 *   on the command line;
 *
 *   operation order is controlled by the order of "if" statements
 *   shown here; the entire sequence can be repeated with optional
 *   pause between each iteration;
 *
 *------------------------------------------------------------------------------------------*/

void manager (struct plc * plc, signed count, signed pause, uint8_t trafficgen_rate)

{

	while (count--)
	{
		if (_anyset (plc->flags, PLC_VERSION))
		{
			VersionInfo2 (plc);
		}
		if (_anyset (plc->flags, PLC_DEV_TRAFFIC))
		{
			Traffic3 (plc, trafficgen_rate);
		}
		if (_anyset (plc->flags, PLC_DEV_TRAFFIC_UNIDI))
		{
			Traffic3_unidi (plc, trafficgen_rate);
		}
		if (_anyset (plc->flags, PLC_NETWORK))
		{
			PhyRates2 (plc);
		}
		if (_anyset (plc->flags, PLC_RESET_DEVICE))
		{
			ResetDevice (plc);
		}
		sleep (pause);
	}
	return;
}

/*====================================================================*
 *
 *   int main (int argc, char const * argv[]);
 *
 *   parse command line, populate plc structure and perform selected
 *   operations; show help summary if asked; see getoptv and putoptv
 *   to understand command line parsing and help summary display; see
 *   plc.h for the definition of struct plc;
 *
 *   the command line accepts local and remote device address and it
 *   performs the specified operations on it; the address order is significant 
 *   but the option order is not; the default address is a local broadcast 
 *   that causes all devices on the local H1 interface to respond but not 
 *   those at the remote end of the powerline;
 *
 *   the default address is 00:B0:52:00:00:01; omitting the address
 *   will automatically address the local device; some options will
 *   cancel themselves if this makes no sense;
 *
 *   the default interface is eth1 because most people use eth0 as
 *   their principle network connection; you can specify another
 *   interface with -i or define environment string PLC to make
 *   that the default interface and save typing;
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv [])

{
	extern struct channel channel;
	static char const * optv [] =
	{
		"a:cd:ei:l:o:nqrRtTuvw:x",
		"node peer [> stdout]",
		"Traffic generation utility for 75xx devices",
		"a\trate(Mbps)[" LITERAL (DEFAULT_TRAFFICGEN_RATE) "] of traffic sent to peer device",
		"c\tdisplay coded PHY rates",
		"d n\ttraffic duration is (n) seconds per leg [" LITERAL (PLC_ECHOTIME) "]\n\t0 to stop the traffic",
		"e\tredirect stderr to stdout",
#if defined (WINPCAP) || defined (LIBPCAP)

		"i n\thost interface is (n) [" LITERAL (CHANNEL_ETHNUMBER) "]",

#else

		"i s\thost interface is (s) [" LITERAL (CHANNEL_ETHDEVICE) "]",

#endif

		"l n\tloop (n) times [" LITERAL (PLCRATE_LOOP) "]",
		"n\tnetwork TX/RX information",
		"o n\tread timeout is (n) milliseconds [" LITERAL (CHANNEL_TIMEOUT) "]",
		"q\tquiet mode",
		"r\trequest device information",
		"R\treset device with VS_RS_DEV",
		"t\tgenerate network traffic (one-to-one(peer-to-node)) ",
		"T\tgenerate network traffic (one-to-one(peer-to-node))uni directional" , 
		"u\tdisplay uncoded PHY rates",
		"v\tverbose mode",
		"w n\twait (n) seconds [" LITERAL (PLCRATE_WAIT) "]",
		"x\texit on error",
		(char const *) (0)
	};

#include "../plc/plc.c"

	uint8_t trafficgen_rate = DEFAULT_TRAFFICGEN_RATE;
	signed loop = PLCRATE_LOOP;
	signed wait = PLCRATE_WAIT;
	signed c;
	optind = 1;
	if (getenv (PLCDEVICE))
	{

#if defined (WINPCAP) || defined (LIBPCAP)

		channel.ifindex = atoi (getenv (PLCDEVICE));

#else

		channel.ifname = strdup (getenv (PLCDEVICE));

#endif

	}
	plc.timer = PLC_ECHOTIME;
	
	while (~ (c = getoptv (argc, argv, optv)))
	{
		switch (c)
		{
		case 'a':
			trafficgen_rate = (unsigned) (uintspec (optarg, 0, 50));
			break;
		case 'c':
			_clrbits (plc.flags, PLC_UNCODED_RATES);
			break;
		case 'd':
			plc.timer = (unsigned) (uintspec (optarg, 0, 60));
			break;
		case 'e':
			dup2 (STDOUT_FILENO, STDERR_FILENO);
			break;
		case 'i':

#if defined (WINPCAP) || defined (LIBPCAP)

			channel.ifindex = atoi (optarg);

#else

			channel.ifname = optarg;

#endif

			break;
		case 'l':
			loop = (unsigned) (uintspec (optarg, 0, UINT_MAX));
			break;
		case 'n':
			_setbits (plc.flags, PLC_NETWORK);
			break;
		case 'o':
			channel.timeout = (signed) (uintspec (optarg, 0, UINT_MAX));
			break;
		case 'q':
			_setbits (plc.flags, PLC_SILENCE);
			break;
		case 'r':
			_setbits (plc.flags, PLC_VERSION);
			break;
		case 'R':
			_setbits (plc.flags, PLC_RESET_DEVICE);
			break;
		case 't':
			_setbits (plc.flags, PLC_DEV_TRAFFIC);
			break; 
		case 'T' :
			_setbits (plc.flags, PLC_DEV_TRAFFIC_UNIDI);
			break; 
		case 'u':
			_setbits (plc.flags, PLC_UNCODED_RATES);
			break;
		case 'v':
			_setbits (channel.flags, CHANNEL_VERBOSE);
			_setbits (plc.flags, PLC_VERBOSE);
			break;
		case 'w':
			wait = (unsigned) (uintspec (optarg, 0, 3600));
			break;
		case 'x':
			_setbits (plc.flags, PLC_BAILOUT);
			break;
		default: 
			break;
		}
	}

	argc -= optind;
	argv += optind;

	if (_allclr (plc.flags, (PLC_VERSION | PLC_DEV_TRAFFIC | PLC_RESET_DEVICE | PLC_DEV_TRAFFIC_UNIDI)) || (optind == 1))
	{
		_setbits (plc.flags, PLC_NETWORK);
	}

	if (_anyset(plc.flags, PLC_DEV_TRAFFIC) || _anyset(plc.flags, PLC_DEV_TRAFFIC_UNIDI))
	{
		if (! argc || ! argv)
		{
			error (1, ECANCELED, "No node address given");
		}

		if (! hexencode (plc.RDA, sizeof (plc.RDA), (char const *) (*argv)))
		{
			error (1, errno, PLC_BAD_MAC, *argv);
		}

		argv++;
		argc--;
	}

	openchannel (& channel);
	if (! (plc.message = malloc (sizeof (* plc.message))))
	{
		error (1, errno, PLC_NOMEMORY);
	}

	if (_anyset(plc.flags, PLC_DEV_TRAFFIC_UNIDI) && (! argc || ! argv))
	{
		error (1, ECANCELED, "No peer address given");
	}

	if (! argc)
	{
		manager (& plc, loop, wait, trafficgen_rate);
	}

	if ((argc) && (* argv))
	{
		if (! hexencode (channel.peer, sizeof (channel.peer), (char const *) (*argv)))
		{
			error (1, errno, PLC_BAD_MAC, * argv);
		}
		manager (& plc, loop, wait, trafficgen_rate);
	}
	free (plc.message);
	closechannel (& channel);
	exit (0);
}

